### PR TITLE
Ensure that the URL and CanonicalRequest use the same url and sorting of query parameters

### PR DESCRIPTION
**Pull request by onno-vos-dev**
The meat of this change is to make sure that URLs such as `https://localhost:443/bucket/key/?list-type=2` with query params: `prefix=foo&delimiter=/` are the same **both both in encoding and sorting** when creating in the CanonicalRequest and the request url that is sent off to hackney. Failing to do so, leads to the dreaded (read: un-debuggable) `SignatureDoesNotMatch` error. 

Rather than patching in two places, I'm now ensuring that anything that comes into `aws_request:split_url/1` is already fully encoded and ready to rock and roll so there's no need for the good old `aws_request:fix_qs/1` as that is now handled by `aws_request:add_query/2` instead. 

Tested change on `list_objects_v2` as well as `list_objects` whereas in particular `list_objects_v2` turned out to be the problematic one. Tried a few random s3 endpoints (which are easiest for me to test) which all worked fine :-) 

I haven't looked at the aws-elixir code yet but my best guess would be that `aws-signature` handles this more gracefully and hence doesn't fail. Pointing to even more reason to migrate to `aws-signature` as part of https://github.com/aws-beam/aws-erlang/issues/63

### CODE REVIEW

File: src/aws_request.erl [hunk 1]
```diff
@@ -79,11 +79,12 @@ build_custom_headers(ParamsCustomHeadersMapping, Params0)
 
 %% @doc Add querystring to url is there are any parameters in the list
 -spec add_query(binary(), [{binary(), any()}]) -> binary().
-add_query(Url, []) ->
-    Url;
-add_query(Url, Query) ->
-    QueryString = aws_util:encode_query(Query),
-    aws_util:binary_join([Url, QueryString], <<"?">>).
+add_query(Url0, Query0) ->
+  Url = hd(string:split(Url0, <<"?">>)),
```
Comment (onno-vos-dev) [inline src/aws_request.erl:83]: Not a big fan of this myself, any suggestions? :)

File: src/aws_util.erl [hunk 2]
```diff
@@ -76,19 +76,22 @@ uri_encode_path_byte(Byte) ->
...
-  uri_string:compose_query(KVs);
+  fix_qs(List);
 encode_query(Map) when is_map(Map) ->
   encode_query(maps:to_list(Map)).
 
+%% When signing a request, the query string for query params that do
+%% no contain a value such as "key" should be encoded as "key=".
+%% Without this fix, the request will result in a SignatureDoesNotMatch error.
+fix_qs(Query) when is_list(Query) ->
+  uri_string:compose_query(
```
Comment (onno-vos-dev) [inline src/aws_util.erl:87]: Fold this into encode_query/1 to make the diff a bit easier

File: src/aws_request.erl [hunk 3]
```diff
@@ -441,28 +432,23 @@ split_url_test() ->
...
     ?assertEqual(
        {<<"/index">>, <<"one=1">>},
        split_url(<<"https://username:secret@example.com:80/index?one=1">>)).
 
-%% split_url/1 returns a patched up binary if a query string with no value is present.
-split_url_with_uri_elements_with_no_value_test() ->
-    ?assertEqual(
-       {<<"/index">>, <<"one=">>},
-       split_url(<<"https://username:secret@example.com:80/index?one">>)).
+add_query_test() ->
```
Comment (onno-vos-dev) [inline src/aws_request.erl:441]: Add more tests

File: src/aws_request.erl (hunk 1 above)
Comment (jadeallenx) [reply src/aws_request.erl:83]: OK. My suggestion would be:

```erlang
     [ BaseUrl | _Rest ] = string:split(Url0, <<"?">>),
     %% ...
```

File: src/aws_request.erl [hunk 4]
```diff
@@ -79,11 +79,12 @@ build_custom_headers(ParamsCustomHeadersMapping, Params0)
...
 %% @doc Add querystring to url is there are any parameters in the list
 -spec add_query(binary(), [{binary(), any()}]) -> binary().
-add_query(Url, []) ->
-    Url;
-add_query(Url, Query) ->
-    QueryString = aws_util:encode_query(Query),
-    aws_util:binary_join([Url, QueryString], <<"?">>).
+add_query(Url0, Query0) ->
+  [Url | _] = string:split(Url0, <<"?">>),
+  HackneyUrl = hackney_url:parse_url(Url0),
```
Comment (anha0825) [inline src/aws_request.erl:84]: hackney_url:parse_url/1 is used to get the query string, but the not the URL. Would something like this work?

```erlang
HackneyUrl = hackney_url:parse_url(Url0),
NewQs = aws_util:encode_query(hackney_url:parse_qs(HackneyUrl#hackney_url.qs) ++ Query0)),
HackneyUrlWithAddedQs = HackneyUrl#hackney_url{qs = NewQs},
hackney_url:unparse_url(HackneyUrlWithAddedQs).
```

**Review by anha0825 — APPROVED** [review]

**Pull request merged**